### ISSUE TITLE
Highlight reconstruction problems

### ISSUE THREAD

**Issue by masc4ii** [ilia3101/MLV-App#90]
@ilia3101 : I am short before going crazy with highlight reconstruction. Last week we solved a problem with dual iso in relation to highlight reconstruction. Today I got some 7D clips (non dual iso) and here highlight reconstruction does not work - not with your algorithm and not with mine. 
@ all: 
- Ilias algorithm looks via transformation matrix which value is theoretical the highest green (before this channel clipped), and reconstructs for this pixels the green value.
- My algorithm searches in the green channel of the histogram of each frame for the brightest peak, adds a small range around this, and reconstructs for this pixels the green value.

For the 7D clips both does not work. I tested a bit around and got the highlights reconstructed with this:
(before) 
```
if (tmp1b == processing->highest_green) //does not work for 7D clips
{...}
```

(after) 
```
if (tmp1b >= processing->highest_green - 1500 && tmp1b <= processing->highest_green)
{...}
```
So I added a negative range around the theoretical highest green. Note: that works only for profile "Tonemapped" - in "Linear" the pink is there again! 😛 

Do you have other ideas how to fix this? Are there completely different algorithms we could try out?

**Comment by masc4ii** [comment]
Some analysis: look how huge the green bar is in waveform scope, and where it changes when correcting the highlights:
![bildschirmfoto 2018-06-13 um 22 33 08](https://user-images.githubusercontent.com/30245296/41376582-ea42fd90-6f59-11e8-90cf-545343a3f668.png)
![bildschirmfoto 2018-06-13 um 22 33 21](https://user-images.githubusercontent.com/30245296/41376595-f2354774-6f59-11e8-9303-86316552c0ff.png)

**Comment by bouncyball-git** [comment]
You know what, I'm just curious why we are reconstructing only green. What if blue or red is clipped?

**Comment by masc4ii** [comment]
> You know what, I'm just curious why we are reconstructing only green. What if blue or red is clipped?

Yes, I also thought about that. But in 99% of the cases I saw the green channel clipping. The blue one clippes e.g. for blue lights, but that does not look so ugly.

> Edit: I never saw better reconstructing than ACR does.

Really?
![bildschirmfoto 2018-06-14 um 17 45 13](https://user-images.githubusercontent.com/30245296/41423193-f5f23132-6ffa-11e8-9937-ae1e99ea6ced.png)

I have another idea: I limited the colors for which the reconstruction is made. So the range can be really huge (see above, e.g. 15000), but I forbid to reconstuct green values - so the trees stay green 😄 - that seems to work, if e.g. no pink flowers are in the picture...

**Comment by masc4ii** [comment]
Ilia wrote to the ML-Forum: 
> highlight reconstruction is for fixing pink highlights, it just has the requirement of a correct white level.

Does that mean, for all clips where it does not work, we just have to adjust the white level manually? Is that possible in the deeps of raw processing in MLV App?

**Comment by masc4ii** [comment]
In https://github.com/ilia3101/MLV-App/commit/be243f9dd6eaefdcac4afef79e4d28e9c07e87db I commited some code for an "aggressive highlight reconstruction mode". This kills all pink in the approx. 1/4 highest green. It works e.g. for the 700D MLV above or for this one: https://www.dropbox.com/s/iy95x9zhupesa4n/M13-1329_7Dpink.MLV?dl=0

**Comment by bouncyball-git** [comment]
We should add white level slider.

**Comment by masc4ii** [comment]
What range should they have? 0..2^14 or more because of dual iso? I hope it is possible to fit the right values... Are you able to set the values in Raw Correction module? I do the Gui and Receipt things...

**Comment by bouncyball-git** [comment]
Yes default slider values should be levels from mlv.

for black and white we can use whole ranges as usual:

10bit 0-1023
12bit 0-4095
14bit 0-16383
16bit 0-65535

**Comment by bouncyball-git** [comment]
Note: 16bit range is not necessary b/c we have to correct original white level 10/12/14bits. Even when clip is lossless there is only 14bits.

**Comment by masc4ii** [comment]
Thanks for the info, that makes it easier again. I'll try out next if the function `processingSetBlackAndWhiteLevel(...)` could do the job already.

**Comment by masc4ii** [comment]
Please try out latest commit. The sliders are there and (hopefully) working. I found out some strange stuff: in most <14bit clips from any camera, the black level and white level is very often not correct. The problem is now, that these wrong values are loaded explicitely. That leads to a black or white picture.
![bildschirmfoto 2018-06-18 um 21 17 27](https://user-images.githubusercontent.com/30245296/41557303-7c938c0a-733d-11e8-9ba1-61f2276bcf04.png)
(This dual iso clip is black before pressing dual iso button. Look at the values - they are wrong.)
@dannephoto : this is one of your 11bit dual iso clips.

**Comment by masc4ii** [comment]
Are the white and black values listed in the box calculated to 14bits? This could explain a lot...

**Comment by dannephoto** [comment]
Dual iso is toooo tricky. Let´s test regular footage first. But heeey. How cool is it with sliders for this. Awesome. Will test tomorrow or when I get some time over here.

**Comment by masc4ii** [comment]
Whaaaaa... I corrected it, that the MLV values are seen in 14bit, but now other files don't work - I have a 10bit file where BL=128 and WL=1013.... hmmmmmmm
So what is right now?!

**Comment by masc4ii** [comment]
Now I got it... so have fun trying it out 😄

**Comment by masc4ii** [comment]
@bouncyball-git : when dual iso is applied to the clip, it seems as if dual iso algorithm is setting up white level (and maybe also black level?). This makes problems between dual iso and the new sliders:
- open clip
- adjust white and black somewhere in the middle (something happens, looks ugly, just for test)
- activate dual iso (looks good)
- minimal adjustment to the 2 sliders (looks ugly again)
- deactivate & activate dual iso (looks good again)
...

Should I deactivate (gray out) the sliders for dual iso?

**Comment by bouncyball-git** [comment]
> So what is right now?!

Everything was right out of the box no need to recalculate anything.

**Comment by dannephoto** [comment]
Oh lawdy lawd. We're in deep pink shit now.

**Comment by masc4ii** [comment]
> Everything was right out of the box no need to recalculate anything.
10bit uncompressed 128-1013
so called 10bit lossless 2047-3817
I wanted you to understand that uncompressed 10bits and the lossless 10bits are very different things!!!

Yes, I think I got that right now. I must use getMlvBitDepth() instead of what we were using in the info box. Then everything is fine and fits.

The question is now only, what do the sliders when dual iso is on - because the shown slider values are wrong after enabling dual iso.

**Comment by bouncyball-git** [comment]
I added set raw black level in the on_horizontalSliderRawBlack_valueChanged (also same for white level) because theoretically if we change raw levels we should set them in the mlv object itself.

It helped a bit for non dual iso but not for dual iso.

**Comment by masc4ii** [comment]
Did you have any problem for non dual iso clips? For me at least this was working 😄

**Comment by bouncyball-git** [comment]
Right resetting will be broken (it was just a quick hack, better use some dedicated var and do not touch the original MLV vals. However this will require some changes here and there.

It seems for non dual iso this works. Changing white level by 1 and then back to original seems to have repeatable effect, not for dual iso though...

**Comment by masc4ii** [comment]
Yes, I also wouldn't touch the original MLV header values, to make everything always repeatable. The only place in the code where something was done with the white and black level is the processing module (for non dual iso part). For dual iso I need more time to have a look where the levels are calculated and where I can grab them to set my sliders 👻 . 

The bad thing will be: I can grab such values AFTER the picture is calculated... -> multithread problem. Or we leave it as it is and just block the sliders for dual iso (highlight reconstruction does not work with white level slider for dual iso anyway, but with my modified algorithm).

**Comment by masc4ii** [comment]
Autsch... I see @bouncyball-git : you wanted to set the levels in the raw_info structure, for a correct dng and mlv export, and for all the corrections, right? That doesn't work indeed with my current try 😞 
Oh my dear... this is a lot of work to do!!!

**Comment by masc4ii** [comment]
@bouncyball-git : please try latest commit. All corrections + mlv + dng export should do their job. MLV header can now be changed, because I made 2 variables in mlv_object to save the original default values.

**Comment by bouncyball-git** [comment]
> you wanted to set the levels in the raw_info structure, for a correct dng and mlv export, and for all the corrections, right?

Yes! but not only for this. 

As you know MLV w/b values are grabbed and used in various places in the code (raw corrections) including dual iso processing... so it code finds already changed raw levels it would be better.

**Comment by bouncyball-git** [comment]
>  MLV header can now be changed, because I made 2 variables in mlv_object to save the original default values.

Cool! Will try.
